nexus_keycode/protocols/channel_origin_commands.py

```diff
@@ -0,0 +1,458 @@
+        :type body: :class:`str`
+        :param auth: arbitrary digits of auth field
+        :type auth: :class:`str`
+        """
+        if not isinstance(type_, OriginCommandType):
+            raise TypeError("Must supply valid OriginCommandType.")
+
+        self.type_code = type_.value
+        self.body = body
+        self.auth = auth
+
+    def __str__(self):
+        return self.to_digits()
+
+    def __repr__(self):
+        return (
+            "{}.{}("
+            "{type_code!r}, "
+            "{body!r}, "
+            "{auth!r},))").format(
+            self.__class__.__module__,
+            self.__class__.__name__,
+            **self.__dict__)
+
+    def to_digits(self, obscured=True):
+        # type: (bool) -> str
+        # String of digits making up this Nexus Channel "Token".
+
+        result = "{}{}{}".format(
+            self.type_code,
+            self.body,
+            self.auth
+        )
+        if obscured:
+            # obscure all digits except MAC/auth
+            obscured_digit_count = len(result) - len(str(self.auth))
+            result = full_obscure(result, obscured_digit_count=obscured_digit_count)
+
+        return result
+
+    @staticmethod
+    def digits_from_siphash(siphash_function, digits=6):
+        """ Return the least-significant digits from a Siphash function.
+
+        Defaults to 6, may be increased.
+        """
+        format_str = "{{:0{}d}}".format(digits)
+        return format_str.format(
+            siphash_function.hash() & 0xffffffff
+        )[-digits:]
+
+
+class GenericControllerActionToken(ChannelOriginCommandToken):
+    """ Not intended to be instantiated directly.
+
+    see: `unlink_all_accessories`.
+    """
+
+    _origin_command_type = OriginCommandType.GENERIC_CONTROLLER_ACTION
+
+    def __init__(
+            self,
+            type_,
+            controller_command,
+            auth
+    ):
+        super(GenericControllerActionToken, self).__init__(
+            type_=self._origin_command_type,
+            body=controller_command,
+            auth=self.digits_from_siphash(auth)
+        )
+
+    @enum.unique
+    class GenericControllerActionType(enum.Enum):
+        """ Types of 'generic controller actions' that are possible.
+        Types 0-20 are reserved for Angaza use. Other types may be 'custom'
+        as needed.
+        """
+
+        # Delete all accessories from the receiving controller
+        UNLINK_ALL_ACCESSORIES = 0
+        # Unlock all accessories linked to the receiving controller
+        UNLOCK_ALL_ACCESSORIES = 1
+        # Types 2-99 undefined
+
+    @classmethod
+    def _generic_controller_action_builder(
+        cls,
+        type_,
+        controller_command_count,
+        controller_sym_key
+    ):
+        # type: (int, str) -> ChannelOriginCommandToken
+        """ Resulting token:
+
+        1-digit Origin Keycode Type ID (0)
+        2-digit "Origin Controller Commmand" (0-99)
```

Comment:
```suggestion
        1-digit OriginCommandType (0)
        2-digit GenericControllerActionType (0-99)
```

Using the names of the classes defined here -- thought this was clearer.